### PR TITLE
[TR] PIM-4629: how to add a custom action in rule engine

### PR DESCRIPTION
**Pull request by solivier**

### CODE REVIEW

File: cookbook/rule/add_custom_action_rule_engine.rst [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+--------------
+
+**This cookbook is about a feature only provided in the Enterprise Edition.**
+
+This cookbook assumes that you already created a new bundle to add your custom rule. Let's assume its namespace is `Acme\\CustomBundle`.
+
+Create a custom action
+----------------------
+
+In this cookbook we are going to see how to add a custom action in the rule engine that allow to concatenate attribute value following a pattern
```
Comment (MarieMinasyan) [inline cookbook/rule/add_custom_action_rule_engine.rst]: that allows to

File: cookbook/rule/add_custom_action_rule_engine.rst [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+
+        /**
+         * @param string $pattern
+         *
+         * @return ProductAddActionInterface
+         */
+        public function setPattern($pattern);
+    }
+
+Then create the object that will extend the interface we created.
```
Comment (MarieMinasyan) [inline cookbook/rule/add_custom_action_rule_engine.rst]: implement

File: cookbook/rule/add_custom_action_rule_engine.rst [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+         */
+        public function setPattern($pattern)
+        {
+            $this->pattern = $pattern;
+
+            return $this;
+        }
+    }
+
+We also need to create a denormalizer that will return our previous object that handle the data. It will convert the array into an object.
```
Comment (MarieMinasyan) [inline cookbook/rule/add_custom_action_rule_engine.rst]: that handles the data

File: cookbook/rule/add_custom_action_rule_engine.rst [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+        /**
+         * {@inheritdoc}
+         */
+        public function supportsDenormalization($data, $type, $format = null)
+        {
+            return isset($data['type']) && ProductPatternActionInterface::ACTION_TYPE === $data['type'];
+        }
+    }
+
+For our example we need to create an `ExistingAttributeValidator` that will check if the attributes provided in the rule file exists or not. It will raise a violation and skip the item in the case the attribute does not exist.
```
Comment (MarieMinasyan) [inline cookbook/rule/add_custom_action_rule_engine.rst]: "attributes provided in the rule file exist" or "attribute provided in the rule file exists"

Comment (MarieMinasyan) [inline cookbook/rule/add_custom_action_rule_engine.rst]: in case if the attribute does not

File: cookbook/rule/add_custom_action_rule_engine.rst [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+        {
+            $this->patternActionClass = $patternActionClass;
+        }
+
+        /**
+         * {@inheritdoc}
+         */
+        public function denormalize($data, $class, $format = null, array $context = [])
+        {
+            return new $this->patternActionClass($data);
```
Comment (ClementGautier) [inline cookbook/rule/add_custom_action_rule_engine.rst]: I think its too heavy for a cookbook, you can simply do:

return new ProductPatternAction($data) and remove the constructor

File: cookbook/rule/add_custom_action_rule_engine.rst [hunk 6]
```diff
@@ -0,0 +1,429 @@
...
+        /**
+         * {@inheritdoc}
+         */
+        public function supports(ActionInterface $action)
+        {
+            return $action instanceof ProductPatternActionInterface;
+        }
+    }
+
+Then we need to create the object that will handle the data, and first we will create the interface for that object:
```
Comment (ClementGautier) [inline cookbook/rule/add_custom_action_rule_engine.rst]: To me, its overkill to do the interface in the cookbook, we should focus the essential.

**Comment by ClementGautier** [comment]
GTM after comment fixed (interface removed and classname not injected to the denormalizer)

**Comment by ClementGautier** [comment]
Wait, another thing : Could we remove the Denormalizer and use `namespace Symfony\Component\Serializer\Normalizer` instead ? It would be easier for reader to understand the meaning of the Action instead of focusing on details. BTW removing this would make the __construct of ProductPatternAction useless and you could be able to remove it.

File: cookbook/rule/add_custom_action_rule_engine.rst (hunk 6 above)
Comment (solivier) [inline cookbook/rule/add_custom_action_rule_engine.rst]: in fact we need it as in the 'Akeneo\Bundle\RuleEngineBundle\Model\Rule::addAction()' we check on the interface Akeneo\Bundle\RuleEngineBundle\Model\ActionInterface

File: cookbook/rule/add_custom_action_rule_engine.rst [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+        }
+
+        /**
+         * {@inheritdoc}
+         */
+        public function setField($field)
+        {
+            $this->field = $field;
+
+            return $this;
```
Comment (ClementGautier) [inline cookbook/rule/add_custom_action_rule_engine.rst]: to remove

File: cookbook/rule/add_custom_action_rule_engine.rst [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+        }
+
+        /**
+         * {@inheritdoc}
+         */
+        public function setOptions(array $options = [])
+        {
+            $this->options = $options;
+
+            return $this;
```
Comment (ClementGautier) [inline cookbook/rule/add_custom_action_rule_engine.rst]: to remove

File: cookbook/rule/add_custom_action_rule_engine.rst [hunk 9]
```diff
@@ -0,0 +1,334 @@
...
+        /**
+         * @param array $attributes
+         *
+         * @return ProductAddActionInterface
+         */
+        public function setAttributes(array $attributes = [])
+        {
+            $this->attributes = $attributes;
+
+            return $this;
```
Comment (ClementGautier) [inline cookbook/rule/add_custom_action_rule_engine.rst]: to remove

File: cookbook/rule/add_custom_action_rule_engine.rst [hunk 10]
```diff
@@ -0,0 +1,334 @@
...
+        /**
+         * @param string $pattern
+         *
+         * @return ProductAddActionInterface
+         */
+        public function setPattern($pattern)
+        {
+            $this->pattern = $pattern;
+
+            return $this;
```
Comment (ClementGautier) [inline cookbook/rule/add_custom_action_rule_engine.rst]: to remove

**Comment by ClementGautier** [comment]
little comments then GTM

**Comment by sumbobyboys** [comment]
GTM

**Pull request merged**